posts/2022-06-27-foldleft-introduction-en.adoc

```diff
@@ -0,0 +1,372 @@
+Left(java.lang.IllegalArgumentException: Unable to delete the task due to status is Enrolled)
+*/
+----
+
+Some details :
+
+[source, scala]
+----
+acc.flatMap { players =>
+            current.map(_ +: players)
+          }
+----
+
+The flatMap allows you to access the Player Sequence located in the Right of the accumulator's Either and return an `Either[Exception, Seq[Player]]` instead of an `Either[Exception, Either[Exception, Seq[Player]]`.
+
+[source, scala]
+----
+.left
+.map(error => ... 
+----
+If there is no Right, then Left is considered a return type. As there is only one possible Left in our return type Either[Exception, Seq[Player]], then processing is stopped as soon as it is filled in.
+
+With the Cats library, the result would be as follows :
+
+[source, scala]
+----
+import cats.data.{EitherT, Validated}
+import cats.implicits._
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContextExecutor, Future}
+implicit val executor: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
+sealed trait PlayerStatus
+object PlayerStatus {
+  case object Available extends PlayerStatus
+  case object Enrolled extends PlayerStatus
+  case object Resting extends PlayerStatus
+  case object Deleted extends PlayerStatus
+}
+case class Player(name: String, currentStatus: PlayerStatus) {
+  def updateStatus(
+      status: PlayerStatus
+  ): Validated[Exception, Player] =
+    if (currentStatus == PlayerStatus.Enrolled)
+      Validated.invalid[Exception, Player](new IllegalArgumentException(s"status is $currentStatus"))
+    else Validated.valid[Exception, Player](copy(currentStatus = status))
+}
+case class Team(players: Seq[Player])
+val team = Team(
+  Seq(
+    Player("player1", PlayerStatus.Available),
+    Player("player2", PlayerStatus.Resting),
+    Player("player3", PlayerStatus.Enrolled) // The status that causes the interruption
+  )
+)
+val resultEitherT: EitherT[Future, IllegalArgumentException, Seq[Player]] =
+  for {
+    updatedPlayers <- EitherT.fromEither[Future] {
+      team.players
+        .map(_.updateStatus(PlayerStatus.Deleted))
+        .foldLeft[Validated[Exception, Seq[Player]]](Validated.Valid(Seq.empty[Player])) { (acc, current) =>
+          acc.andThen { players =>
+            current.map(_ +: players)
+          }
+        }
+        .leftMap(error => new IllegalArgumentException(s"Unable to delete the task due to ${error.getMessage}"))
+        .toEither
+    }
+  } yield updatedPlayers
+Await.result(resultEitherT.value, 1.second)
+/* Result =>
+Left(java.lang.IllegalArgumentException: Unable to delete the task due to status is Enrolled)
+*/
+----
+
+Précisions :
+
+[source, scala]
+----
+acc.andThen { players =>
+            current.map(_ +: players)
+          }
+----
+
+Again, this is the Cats version of left.map(...)
+Finally, we wrap our block with `EitherT.fromEither[Future] { { ... }.toEither }` to change the type from `Validated` to `EitherT`.
```

Comment:
This is the Cats version of left.map(...)

Finally, we wrap our block with `EitherT.fromEither[Future] { { ... }.toEither }` to change the type from `Validated` to `EitherT`.

Note that we have used `EitherT` and `Validated`, two Cats specific types.